### ISSUE TITLE
Fix and improve OriginValidator

### ISSUE THREAD

**Issue by bluetech** [django/channels#929]
> Your OS and runtime environment, and browser if applicable

Linux, Firefox.

> The versions of Channels, Daphne, Django, Twisted, and your ASGI backend (channels_redis normally)

- Channels 2.0.2 
- Django 2.0.2

> How you're running Channels (runserver? daphne/runworker? Nginx/Apache in front?)

All of the above.

> What you expected to happen vs. what actually happened

Regarding this example https://github.com/django/channels/blob/d6643d96dd59e153cda13681a6d5a8963b9d577c/docs/topics/security.rst#websockets

- The correct syntax for subdomain wildcard is `.goodsite.com` not `*.goodsite.com` ([reference](https://docs.djangoproject.com/en/2.0/ref/settings/#allowed-hosts)).

- `.goodsite.com` includes `goodsite.com` so `goodsite.com` is redundant.

- It should perhaps be mentioned that `*` is a special value that allows all.

---

Regarding `OriginValidator` itself, why does it only validate the domain rather than the full origin? (An [Origin](https://tools.ietf.org/html/rfc6454) is `scheme://domain:port`).

In terms of the web security model, the following are all different origins:

```
http://example.com:80
http://example.com:8000
https://example.com:443
https://example.com:8443
```

and it should be possible to allow any subset of those rather than all or none.

I know that Django's `ALLOWED_HOSTS` disregards the port (a [Host](https://tools.ietf.org/html/rfc7230#section-5.4) is `domain:port`, no scheme). I don't really agree with that decision, however I don't think it applies to Origins as well: `ALLOWED_HOSTS` can be seen as just an additional protection layer over what the webserver should be doing already; but the list of allowed origins is an explicit policy decision unrelated to the webserver, namely, which 3rd party web sites I want to allow to connect to my WebSocket server.

**Comment by andrewgodwin** [comment]
The wildcard matching is a good point - we should update the code to match Django's implementation, rather than keeping this apparently-broken one from Channels 1. That's going to have to be done in a backwards-compatible way now, though, which I think is fine.

The problem of including the port is a good one. I think the right thing to do is add a mode where it doesn't care about the port for the one that reads from the settings file, and let you be more specific manually if you want.

The scheme should also be something you can probably determine on (to only allow connections from SSL pages), but again optionally, in this case if you provide it we enforce it.

**Comment by andrewgodwin** [comment]
Was fixed in #940
